File: packages/expo-updates/ios/EXUpdates/SelectionPolicy/LauncherSelectionPolicyFilterAware.swift
//  Copyright © 2019 650 Industries. All rights reserved.

import Foundation

/**
 * A LauncherSelectionPolicy which chooses an update to launch based on the manifest
 * filters provided by the server. If multiple updates meet the criteria, the newest one (using
 * `commitTime` for ordering) is chosen, but the manifest filters are always taken into account
 * before the `commitTime`.
 */
@objc(EXUpdatesLauncherSelectionPolicyFilterAware)
@objcMembers
public final class LauncherSelectionPolicyFilterAware: NSObject, LauncherSelectionPolicy {
  let runtimeVersion: String
  private let filterByChannel: Bool
  private let config: UpdatesConfig?

  public required init(runtimeVersion: String, filterByChannel: Bool = false, config: UpdatesConfig? = nil) {
    self.runtimeVersion = runtimeVersion
    self.filterByChannel = filterByChannel
    self.config = config
  }

  public func launchableUpdate(fromUpdates updates: [Update], filters: [String: Any]?) -> Update? {
    let matchingUpdates = updates
      .filter { runtimeVersion == $0.runtimeVersion && SelectionPolicies.doesUpdate($0, matchFilters: filters) }
      .filter { update in
        guard let config = config else {
          return true
        }
        return (update.url == nil && update.requestHeaders == nil) || (update.url == config.updateUrl && update.requestHeaders == config.requestHeaders)
      }

    if filterByChannel, let configuredChannelName = channelNameFromConfig(), !configuredChannelName.isEmpty {
      NSLog("[ChannelSwitch] Filtering updates for channel %@", configuredChannelName)
      let channelFilteredUpdates = matchingUpdates.filter { update in
        guard let updateChannel = channelName(for: update) else {
          return false
        }
        return updateChannel == configuredChannelName
      }

      if !channelFilteredUpdates.isEmpty {
        NSLog("[ChannelSwitch] Found %d updates matching channel %@", channelFilteredUpdates.count, configuredChannelName)
        return channelFilteredUpdates.sorted { $0.commitTime > $1.commitTime }.first
      }

      NSLog("[ChannelSwitch] No updates matched channel %@; falling back to unfiltered selection", configuredChannelName)
    }

    return matchingUpdates.sorted { $0.commitTime > $1.commitTime }.first
  }

  private func channelNameFromConfig() -> String? {
    guard let requestHeaders = config?.requestHeaders else {
      return nil
    }
    return requestHeaders.first(where: { $0.key.lowercased() == "expo-channel-name" })?.value
  }

  private func channelName(for update: Update) -> String? {
    let rawManifest = update.manifest.rawManifestJSON()
    if let branch = rawManifest["branch"] as? String {
      return branch
    }
    if let extra = rawManifest["extra"] as? [String: Any],
      let expoClient = extra["expoClient"] as? [String: Any],
      let expoClientExtra = expoClient["extra"] as? [String: Any],
      let latitudeStage = expoClientExtra["LATITUDE_RELEASE_STAGE"] as? String {
      return latitudeStage
    }
    return nil
  }
}

File: packages/expo-updates/android/src/main/java/expo/modules/updates/selectionpolicy/LauncherSelectionPolicyFilterAware.kt
package expo.modules.updates.selectionpolicy

import android.util.Log
import expo.modules.updates.UpdatesConfiguration
import expo.modules.updates.db.entity.UpdateEntity
import org.json.JSONObject

/**
 * LauncherSelectionPolicy which chooses an update to launch based on the manifest filters
 * provided by the server. If multiple updates meet the criteria, the newest one (using `commitTime`
 * for ordering) is chosen, but the manifest filters are always taken into account before the
 * `commitTime`.
 */
class LauncherSelectionPolicyFilterAware(
  private val runtimeVersion: String,
  private val filterByChannel: Boolean = false,
  private val config: UpdatesConfiguration? = null
) : LauncherSelectionPolicy {

  override fun selectUpdateToLaunch(
    updates: List<UpdateEntity>,
    filters: JSONObject?
  ): UpdateEntity? =
    run {
      val matchingUpdates = updates
        .filter { runtimeVersion == it.runtimeVersion && SelectionPolicies.matchesFilters(it, filters) }
        .filter { update ->
          val config = config ?: return@filter true
          (update.url == null && update.requestHeaders == null) || (update.url == config.updateUrl && update.requestHeaders == config.requestHeaders)
        }

      val channelName = channelNameFromConfig()
      if (filterByChannel && !channelName.isNullOrBlank()) {
        Log.d(TAG, "[ChannelSwitch] Filtering updates for channel $channelName")
        val channelFilteredUpdates = matchingUpdates.filter { update ->
          val updateChannel = channelNameForUpdate(update)
          updateChannel != null && updateChannel == channelName
        }
        if (channelFilteredUpdates.isNotEmpty()) {
          Log.d(TAG, "[ChannelSwitch] Found ${channelFilteredUpdates.size} updates matching channel $channelName")
          return@run channelFilteredUpdates.maxByOrNull { it.commitTime }
        }
        Log.d(TAG, "[ChannelSwitch] No updates matched channel $channelName; falling back to unfiltered selection")
      }

      matchingUpdates.maxByOrNull { it.commitTime }
    }

  private fun channelNameFromConfig(): String? {
    val requestHeaders = config?.requestHeaders ?: return null
    return requestHeaders.entries.firstOrNull { it.key.equals("expo-channel-name", ignoreCase = true) }?.value
  }

  private fun channelNameForUpdate(update: UpdateEntity): String? {
    val manifest = update.manifest
    if (manifest.has("branch")) {
      val branch = manifest.optString("branch", null)
      if (!branch.isNullOrEmpty()) {
        return branch
      }
    }
    val extra = manifest.optJSONObject("extra")
    val expoClient = extra?.optJSONObject("expoClient")
    val expoClientExtra = expoClient?.optJSONObject("extra")
    if (expoClientExtra?.has("LATITUDE_RELEASE_STAGE") == true) {
      val stage = expoClientExtra.optString("LATITUDE_RELEASE_STAGE", null)
      if (!stage.isNullOrEmpty()) {
        return stage
      }
    }
    return null
  }

  private companion object {
    private const val TAG = "ExpoUpdates"
  }
}
